# Marketing Strategy Analysis


## 1. Introduction

Our goal with this project will be to give the most accurate response to the following questions:



*   What is the impact of each marketing strategy and sales visit on Sales (Amount Collected)?
*   Is the same strategy valid for all the different client types?




## 2. Data Loading and Quality Checks


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import scipy.stats

In [2]:
# Way to upload data easily from local to collab
#from google.colab import files
#uploaded = files.upload()

In [3]:
# Being the file already uploaded to drive, the usual path is way faster
data = pd.read_csv("/content/drive/MyDrive/Proyectos ML/DataSets/Campaign-Data.csv")

In [4]:
data.columns

Index(['Client ID', 'Client Type', 'Number of Customers', 'Montly Target',
       'Zip Code', 'Calendardate', 'Amount Collected', 'Unit Sold',
       'Campaign (Email)', 'Campaign (Flyer)', 'Campaign (Phone)',
       'Sales Contact 1', 'Sales Contact 2', 'Sales Contact 3',
       'Sales Contact 4', 'Sales Contact 5', 'Number of Competition'],
      dtype='object')

In [5]:
data.head()

Client ID      Client Type  Number of Customers  Montly Target  Zip Code  \
0  ID-987275  Medium Facility                 2800            125      1003   
1  ID-987275  Medium Facility                 2800            125      1003   
2  ID-987275  Medium Facility                 2800            125      1003   
3  ID-987275  Medium Facility                 2800            125      1003   
4  ID-987275  Medium Facility                 2800            125      1003   

  Calendardate  Amount Collected  Unit Sold  Campaign (Email)  \
0   16-01-2014                 0          0               0.0   
1   16-02-2014           3409460         24               0.0   
2   18-03-2014          10228384         75               0.0   
3   18-04-2014          17047304        123               0.0   
4   19-05-2014          23866224        171               0.0   

   Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  Sales Contact 2  \
0               0.0               0.0              0.0              0.0   
1               0.0               0.0              0.0              0.0   
2               0.0               0.0              0.0              0.0   
3               0.0               0.0              0.0        3547500.0   
4               0.0               0.0              0.0              0.0   

   Sales Contact 3  Sales Contact 4  Sales Contact 5 Number of Competition  
0              0.0              0.0              0.0                   Low  
1              0.0              0.0         322500.0                   Low  
2              0.0              0.0              0.0                   Low  
3        1290000.0              0.0              0.0                   Low  
4              0.0              0.0              0.0                   Low

The **Amount Collected** represents a target variable that we are going to use to help to define our marketing strategy. We are going to use the **Campaign (XXX)** and **Sales Contact X** variables as our independent variables to formulate that marketing strategy per **Client Type**.

## 3. Feature Additions and Engineering

In [6]:
# First we are going tweak the Calendardate variable so it changes from a string into a pandas date object
data["Calendardate"] = pd.to_datetime(data["Calendardate"], dayfirst=True)

In [7]:
# Know we use that previous trick to create two new collumns, one for the month and another for the year
data["Calendar_Month"] = data["Calendardate"].dt.month
data["Calendar_Year"] = data["Calendardate"].dt.year

## 4. Exploratory Data Analysis and Statistical Analysis


### 4.1 Impact of Marketing Strategy on Sales

#### Understanding of Distributions

In [8]:
# First of all, we can see by analyzing the Client Type attribute that most of our clients are large facilities
# then small and lastly medium and private facilities.
data["Client Type"].value_counts(normalize=True)

Client Type
Large Facility      0.459677
Small Facility      0.282258
Medium Facility     0.169355
Private Facility    0.088710
Name: proportion, dtype: float64

In [9]:
# We can continue with our categorical analysis by doing a crosstab (wich is a symple cross tabulation),
# where we are looking at two categories wich are Client Type and Number of Competition.
# It shows that the clients with high competition are the least.
pd.crosstab(data["Number of Competition"], data["Client Type"], margins=True, normalize="columns")

Client Type            Large Facility   Medium Facility  Private Facility  \
Number of Competition                                                       
High                          0.166667         0.166667          0.166667   
Low                           0.833333         0.833333          0.833333   

Client Type            Small Facility        All  
Number of Competition                             
High                          0.166667  0.166667  
Low                           0.833333  0.833333

In [10]:
# Now, by grouping by the Number of Competition, we can proceed to calculate the mean for each
# column for each value the variable takes.
data.groupby('Number of Competition').mean(numeric_only=True)

Number of Customers  Montly Target  Zip Code  \
Number of Competition                                                 
High                           1456.935484      75.080645    1003.0   
Low                            1456.935484      75.080645    1003.0   

                       Amount Collected   Unit Sold  Campaign (Email)  \
Number of Competition                                                   
High                       2.974789e+07  213.127016     105398.938508   
Low                        1.445570e+07  103.132258     150862.165766   

                       Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  \
Number of Competition                                                        
High                      994046.717540      45198.036895    146945.564516   
Low                       623692.979839      26693.304194    128219.758065   

                       Sales Contact 2  Sales Contact 3  Sales Contact 4  \
Number of Competition                                                      
High                      2.685333e+06     1.786754e+06     72172.379032   
Low                       1.890916e+06     1.883634e+06     70481.854839   

                       Sales Contact 5  Calendar_Month  Calendar_Year  
Number of Competition                                                  
High                       8452.620968            10.5         2015.0  
Low                       15864.919355             5.7         2014.4

By analyzing these results we can see that the Amount Collected for High Competition is more than double than Low Competition.

With both last results we get that even the majority of the clients are in a Low Competition situation, in terms of sells, clients with High Competition are outperforming them by far.

Also we see how High Competition clients atract half of the customers, while they only represent roughly a 17% of the clients.

In [11]:
# Now we group by the client type and calculate the mean again
data.groupby('Client Type').mean(numeric_only=True)

Number of Customers  Montly Target  Zip Code  \
Client Type                                                      
Large Facility            1380.842105      71.578947    1003.0   
Medium Facility           3940.761905     202.857143    1003.0   
Private Facility           400.727273      20.454545    1003.0   
Small Facility             422.514286      21.285714    1003.0   

                  Amount Collected   Unit Sold  Campaign (Email)  \
Client Type                                                        
Large Facility        1.999880e+07  143.098684     142273.609649   
Medium Facility       4.075997e+07  290.583333     437217.097817   
Private Facility      5.030246e+06   35.784091       5183.715152   
Small Facility        1.637759e+06   11.689286      11975.986310   

                  Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  \
Client Type                                                             
Large Facility        8.192056e+05      45595.436623    133667.763158   
Medium Facility       1.552603e+06      49176.847619    398645.833333   
Private Facility      2.272919e+05       5522.470455      1221.590909   
Small Facility        9.120875e+04          0.000000      8062.500000   

                  Sales Contact 2  Sales Contact 3  Sales Contact 4  \
Client Type                                                           
Large Facility       2.034013e+06     2.017039e+06    119287.280702   
Medium Facility      4.822783e+06     4.698646e+06     85104.166667   
Private Facility     6.376705e+05     4.434375e+05      3664.772727   
Small Facility       7.617143e+05     3.727946e+05      4223.214286   

                  Sales Contact 5  Calendar_Month  Calendar_Year  
Client Type                                                       
Large Facility       16266.447368             6.5         2014.5  
Medium Facility      33273.809524             6.5         2014.5  
Private Facility     12215.909091             6.5         2014.5  
Small Facility        1535.714286             6.5         2014.5

Here we can see that even tough Medium Facility is not the number one in terms of Client Type of our clients in the database, we see that they are bringing in a lot more money than the other client types.

Taking this in mind, we surely want to study what is the marketing strategy followed in these Medium Facility clients and study if it can be applied to other markets.

#### Correlation Analysis

In [12]:
# Now let's start with a correlation analysis by calculating the correlations between
# the Amount Collected variable and the others
data.corr(numeric_only=True)["Amount Collected"]

Number of Customers    0.607496
Montly Target          0.608204
Zip Code                    NaN
Amount Collected       1.000000
Unit Sold              0.997515
Campaign (Email)       0.248235
Campaign (Flyer)       0.444337
Campaign (Phone)       0.034858
Sales Contact 1        0.277478
Sales Contact 2        0.552112
Sales Contact 3        0.357887
Sales Contact 4        0.236165
Sales Contact 5        0.095795
Calendar_Month         0.139425
Calendar_Year          0.286194
Name: Amount Collected, dtype: float64

In [13]:
## Consolidating Strategy for Targeting

cm = sns.light_palette("green", as_cmap=True)
# Let's create calculate the correlation matrix just for the data that's interesting for the marketing campaing
marketing_corr =  pd.DataFrame(data[["Amount Collected",
                                     "Campaign (Email)",
                                     "Campaign (Flyer)",
                                     "Campaign (Phone)",
                                     "Sales Contact 1",
                                     "Sales Contact 2",
                                     "Sales Contact 3",
                                     "Sales Contact 4",
                                     "Sales Contact 5"]].corr(numeric_only=True)["Amount Collected"]).reset_index()

marketing_corr.columns = ['Impacting Variable', 'Degree of Linear Impact (Correlation)']
marketing_corr = marketing_corr[marketing_corr["Impacting Variable"] != "Amount Collected"]
marketing_corr = marketing_corr.sort_values('Degree of Linear Impact (Correlation)', ascending=False)
marketing_corr.style.background_gradient(cmap=cm)#.set_precision(2)

This visual shows us how marketing strategy variables are correlated with the Amount Collected one, in a descending order.

But what we really want is to see these same correlations broken up by Client Type, because that's the key question for devoloping a more accurate strategy for every client.

In [14]:
## Consolidating Strategy for Targeting

cm = sns.light_palette("green", as_cmap=True)
# Let's create calculate the correlation matrix just for the data that's interesting for the marketing campaing
marketing_corr =  pd.DataFrame(data.groupby('Client Type')[["Amount Collected",
                                     "Campaign (Email)",
                                     "Campaign (Flyer)",
                                     "Campaign (Phone)",
                                     "Sales Contact 1",
                                     "Sales Contact 2",
                                     "Sales Contact 3",
                                     "Sales Contact 4",
                                     "Sales Contact 5"]].corr(numeric_only=True)["Amount Collected"]).reset_index()

marketing_corr = marketing_corr.sort_values(['Client Type', 'Amount Collected'], ascending=False)
marketing_corr.columns = ['Client Type','Variable Impact on Sales', 'Impact']
marketing_corr = marketing_corr[marketing_corr["Variable Impact on Sales"] != "Amount Collected"]
marketing_corr.style.background_gradient(cmap=cm)#.set_precision(2)

With this new visual we can see how each independent variable affects very differently the variable we want to model.

#### Regression Analysis (Market Sales and Strategies)

Let's continue to look at causation by moving away from correlation and building a linear regression model.


In [15]:
# First of all we change the names of the variables to make them easier to access
data.columns=[mystring.replace(" ", "_") for mystring in data.columns]
data.columns=[mystring.replace("(", "") for mystring in data.columns]
data.columns=[mystring.replace(")", "") for mystring in data.columns]

In [16]:
# Now we use the library statsmodels to calculare a linear regression to estimate Amount_Collected variable
# using the marketing campaign related independent variables for the regression.
import statsmodels.api as sm
import statsmodels.formula.api as smf

mod = smf.ols("Amount_Collected ~ Campaign_Email + Campaign_Flyer + Campaign_Phone +\
          Sales_Contact_1 + Sales_Contact_2 + Sales_Contact_3 + Sales_Contact_4 + Sales_Contact_5", data=data)
results = mod.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:       Amount_Collected   R-squared:                       0.480
Model:                            OLS   Adj. R-squared:                  0.478
Method:                 Least Squares   F-statistic:                     342.1
Date:                Tue, 16 Apr 2024   Prob (F-statistic):               0.00
Time:                        09:48:49   Log-Likelihood:                -54512.
No. Observations:                2976   AIC:                         1.090e+05
Df Residuals:                    2967   BIC:                         1.091e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept        1.481e+06   5.12e+05     

What we have just done is nothing different from a classic linear regression:

$$
Y = X*m + b
$$

Or in our case something of the following form:

$$
Y = X_1 \cdot m_1 + X_2 \cdot m_2 + ... + b
$$

Looking at the results of the predictions made by the linear regression model we can see all the campaign tactics with the coefficients attached to them (the ms of the formula), giving us an estimation of the impact every variable has over the dependent variable.

Also by noticing the p-value column (the one with P>|t| as header) we get an estimation of how probable is that the negation of the base hypothesis (the one that states that a certain variable has a signifficant impact over the value of the predicted variable) is true. Taking that in mind we can assert that those variables with a high p-value are the ones whose values don't affect much the results.

Having this in mind now we can proceed by eliminating from our stimations those variables that don't give us important information about the Amount_Collected variable.

For this filtering we'll eliminate from the stimator all the independent variables that have a p-value bigger than 0.05,this is the same as keeping all those independent variables that we can affirm with a 95 percent confident that have an signifficant impact over Amount_collected.

In [17]:
df = pd.read_html(results.summary().tables[1].as_html(), header=0, index_col = 0)[0]

In [18]:
df = df.reset_index()
df = df[df["P>|t|"] < 0.05][["index", "coef"]]
df

index          coef
0        Intercept  1.481000e+06
2   Campaign_Flyer  3.337600e+00
4  Sales_Contact_1  4.236800e+00
5  Sales_Contact_2  3.638200e+00
6  Sales_Contact_3  2.343200e+00
7  Sales_Contact_4  1.094780e+01

#### Regression Analysis (Market Sales and Strategies) - Broken for different account types

In [40]:
consolidated_summary = pd.DataFrame()
for clienttype in list(set(list(data["Client_Type"]))):
  temp_data = data[data["Client_Type"] == clienttype].copy()

  # Stablishing the linear regression model
  mod = smf.ols("Amount_Collected ~ Campaign_Email + Campaign_Flyer + Campaign_Phone +\
          Sales_Contact_1 + Sales_Contact_2 + Sales_Contact_3 + Sales_Contact_4 + Sales_Contact_5", data=temp_data)
  results = mod.fit()

  # Filtering the non-relevant variables
  df = pd.read_html(results.summary().tables[1].as_html(), header=0, index_col = 0)[0]
  df = df.reset_index()
  df = df[df["P>|t|"] < 0.05][["index", "coef"]]

  df.columns = ["Variable", "Coefficient (Impact)"]
  df["Client Type"] = clienttype
  df = df.sort_values("Coefficient (Impact)", ascending = False)
  df = df[df["Variable"] != "Intercept"]

  print(clienttype)
  #consolidated_summary = consolidated_summary.append(df)
  consolidated_summary = pd.concat([consolidated_summary, df], ignore_index=True)
  print(df)

Large Facility 
          Variable  Coefficient (Impact)      Client Type
4  Sales_Contact_1               11.6731  Large Facility 
7  Sales_Contact_4               10.6145  Large Facility 
5  Sales_Contact_2                4.0031  Large Facility 
2   Campaign_Flyer                2.7204  Large Facility 
6  Sales_Contact_3                2.0316  Large Facility 
3   Campaign_Phone               -3.5361  Large Facility 
Small Facility 
          Variable  Coefficient (Impact)      Client Type
5  Sales_Contact_2              0.810100  Small Facility 
3   Campaign_Phone             -0.000003  Small Facility 
Medium Facility
          Variable  Coefficient (Impact)      Client Type
2   Campaign_Flyer                4.1059  Medium Facility
5  Sales_Contact_2                3.5778  Medium Facility
4  Sales_Contact_1                3.1365  Medium Facility
6  Sales_Contact_3                2.1174  Medium Facility
Private Facility
          Variable  Coefficient (Impact)       Client Type
5  Sal

Applying all previous knowledge we get the table sawn just above.

In that table we have a breakdown of each one of our client types and the particular marketing tactic with the coefficient attached to it. For example, we can see how while Campaign_Flyer is the most important variable for Medium Facility clients, it's not as important as some of the Sales_Contacs variables for the Large Facilities.

We have also created a dataframe to store the information of the table, so we'll be able to use it for further analysis.

## 5. Final Recommendations

Now we can use the coefficients stored in the dataframe to see how much return we can derive from every dollar we spend in marketing, here we can clearly see that for different client types the various campaigns and sale contacts are effective with different extend.

**Case Explanation - Medium Facility**

For example, clients with Medium Facility show decent results with Flier Campaigns, having a 4 dollars revenue on average for each dollar spent. Sales Contact 2 is also higly effective, closely followed by Sales Contact 1 and 3 respectively. For all the other marketing strategies,there is not enough evidence that indicates that they have some impact over the revenue, so the investment in them can be dropped to 0.


In [31]:
consolidated_summary

Variable  Coefficient (Impact)       Client Type
0   Sales_Contact_1             11.673100   Large Facility 
1   Sales_Contact_4             10.614500   Large Facility 
2   Sales_Contact_2              4.003100   Large Facility 
3    Campaign_Flyer              2.720400   Large Facility 
4   Sales_Contact_3              2.031600   Large Facility 
5    Campaign_Phone             -3.536100   Large Facility 
6   Sales_Contact_2              0.810100   Small Facility 
7    Campaign_Phone             -0.000003   Small Facility 
8    Campaign_Flyer              4.105900   Medium Facility
9   Sales_Contact_2              3.577800   Medium Facility
10  Sales_Contact_1              3.136500   Medium Facility
11  Sales_Contact_3              2.117400   Medium Facility
12  Sales_Contact_2              6.622300  Private Facility

Let's enhance the table so it's more visually appealling and easy to read.

In [32]:
consolidated_summary.reset_index(inplace = True)
consolidated_summary.drop('index', inplace = True, axis = 1)

In [41]:
consolidated_summary.columns = ["Variable", "Return on Investment", "Account Type"]
consolidated_summary["Return on Investment"] = consolidated_summary["Return on Investment"].apply(lambda x: round(x, 1))
consolidated_summary.style.background_gradient(cmap = "RdYlGn")

In [42]:
def format(x):
  return "${:.1f}".format(x)
consolidated_summary['Return on Investment'] = consolidated_summary['Return on Investment'].apply(format)

In [43]:
consolidated_summary

Variable Return on Investment      Account Type
0   Sales_Contact_1                $11.7   Large Facility 
1   Sales_Contact_4                $10.6   Large Facility 
2   Sales_Contact_2                 $4.0   Large Facility 
3    Campaign_Flyer                 $2.7   Large Facility 
4   Sales_Contact_3                 $2.0   Large Facility 
5    Campaign_Phone                $-3.5   Large Facility 
6   Sales_Contact_2                 $0.8   Small Facility 
7    Campaign_Phone                $-0.0   Small Facility 
8    Campaign_Flyer                 $4.1   Medium Facility
9   Sales_Contact_2                 $3.6   Medium Facility
10  Sales_Contact_1                 $3.1   Medium Facility
11  Sales_Contact_3                 $2.1   Medium Facility
12  Sales_Contact_2                 $6.6  Private Facility

What we end up here is a table for our marketing tactics that allows us to create a campaign.

For example, for a Private Facility the Sales Contact 2 will be the only strategy to follow. On the other hand, when dealing with a client with a Large Facility we will have to pay more attention to Sales Contacts 1 and 4, while not forgetting about Sales Contact 2.

We can also see that for Large Facilities Phone Campaings are negatively correlated with the revenue, what indicates that for every dollar we spend in this type of marketing we are loosing another $3.5, being a really counter-productive strategy.

## 6. Conclusions

### Where Should we Invest More?


**Growing Sales**

**(In-organic Revenue)**

From the analysis, we see that many of the factors correlate with the accounts collected. The tactics that were able to drive the most significant return on investment were the sales contacts. Placing more investment in these areas will benefit an overall campaign.

This **visits might be more significant due to position in the sales cycle**. However, flyers were also important in accounts collected. **Synergies between these tactics may need exploration**.
